src/service/cid-api.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getCookie, setCookie} from '../cookies';
+import {Services} from '../services';
+import {dev} from '../log';
+import {dict} from '../utils/object';
+
+const GOOGLE_API_URL = 'https://ampcid.google.com/v1/publisher:getClientId?key=';
+const API_KEYS = {
+  'googleanalytics': 'AIzaSyA65lEHUEizIsNtlbNo-l2K18dT680nsaM',
+};
+
+const TAG = 'GoogleCidApi';
+const AMP_TOKEN = 'AMP_TOKEN';
+
+/** @enum {string} */
+const TokenStatus = {
+  RETRIEVING: '$RETRIEVING',
+  OPT_OUT: '$OPT_OUT',
+  ERROR: '$ERROR',
+};
+
+const TIMEOUT = 30000;
+const DAY = 24 * 60 * 60 * 1000;
+const YEAR = 365 * DAY;
+
+/**
+ * Client impl for Google CID API
+ */
+export class GoogleCidApi {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    this.win_ = win;
+    /**
+     * @private {!./timer-impl.Timer}
+     */
+    this.timer_ = Services.timerFor(this.win_);
+
+    /**
+     * @private {!Object<string, !Promise<?string>>}
+     */
+    this.cidPromise_ = {};
+  }
+
+  /**
+   * @param {string} scope
+   * @param {string} apiClient
+   * @return {!Promise<?string>}
+   */
+  getScopedCid(scope, apiClient) {
+    const url = this.getUrl_(apiClient);
+    if (!url) {
+      return Promise.resolve(/** @type {?string} */(null));
+    }
+
+    if (this.cidPromise_[scope]) {
+      return this.cidPromise_[scope];
+    }
+    let token;
+    // Block the request if a previous request is on flight
+    // Poll every 200ms. Longer interval means longer latency for the 2nd CID.
+    return this.cidPromise_[scope] = this.timer_.poll(200, () => {
+      token = getCookie(this.win_, AMP_TOKEN);
+      return token !== TokenStatus.RETRIEVING;
+    }).then(() => {
+      if (token === TokenStatus.OPT_OUT) {
+        return null;
+      }
+      if (token === TokenStatus.ERROR) {
+        return getCookie(this.win_, scope);
+      }
```

Comment:
one more thing, to be future-proof,  Should we check and not send the token value in the response when it starts with '$'? I am worried about one of the clients introducing a value that is not understood by other clients. What say? 

Obviously this is not super urgent.